### PR TITLE
pilot: fix issue with TLS and TCP order dependency

### PR DESCRIPTION
**Pull request by howardjohn**
Currently, tls_inspector is only added if its the first Service. If its
after TCP, it is missed.

example config:

```yaml
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  name: tcp
spec:
  addresses:
  - 10.10.10.10/24
  exportTo:
  - .
  hosts:
  - '*.tcp'
  ports:
  - name: tcp-443
    number: 443
    protocol: TCP
---
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  name: tls
spec:
  exportTo:
  - .
  hosts:
  - tls.example.com
  location: MESH_EXTERNAL
  ports:
  - name: https-443
    number: 443
    protocol: HTTPS
  resolution: DNS
```

depending on the order of creation it will work/not work.

**Please provide a description of this PR:**

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
When did this bug occur?

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 1]
```diff
@@ -1009,10 +1010,12 @@ func (lb *ListenerBuilder) buildSidecarOutboundListenerForPortOrUDS(listenerOpts
 		// Merge two TCP filter chains. HTTP filter chain will not conflict with TCP filter chain because HTTP filter chain match for
 		// HTTP filter chain is different from TCP filter chain's.
 		currentListenerEntry.listener.FilterChains = mergeTCPFilterChains(mutable.Listener.FilterChains, listenerOpts, listenerMapKey, listenerMap)
+		currentListenerEntry.listener.ListenerFilters = mergeListenerFilters(currentListenerEntry.listener.ListenerFilters, mutable.Listener.ListenerFilters)
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/listener.go:1013]: The example is tcp over tcp, could not understand why merge listener filter for NoConflict and `TCPOverAuto`

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 2]
```diff
@@ -1539,6 +1542,29 @@ func isConflictWithWellKnownPort(incoming, existing protocol.Instance, conflict
 	return true
 }
 
+func mergeListenerFilters(a, b []*listener.ListenerFilter) []*listener.ListenerFilter {
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/listener.go:1545]: similar to appendListenerFiltersm what's the diff?

Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/listener.go:1545]: I think the target of this function is  to return a union b, if so the implement is not correct

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/listener.go:1545]: Why is it not correct?

**Comment by howardjohn** [comment]
> When did this bug occur?

There is an example of 2 SE in the PR description

**Comment by hzxuzhonghu** [comment]
I am asking to know if this is a regression

**Review by ericvn — APPROVED** [review]
Release note looks good.

**Comment by howardjohn** [comment]
@hzxuzhonghu AFAIK its been broken forever

File: pilot/pkg/networking/core/v1alpha3/listener.go (hunk 2 above)
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/listener.go:1545]: Can you elaborate?

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 3]
```diff
@@ -1009,10 +1010,12 @@ func (lb *ListenerBuilder) buildSidecarOutboundListenerForPortOrUDS(listenerOpts
 		// Merge two TCP filter chains. HTTP filter chain will not conflict with TCP filter chain because HTTP filter chain match for
 		// HTTP filter chain is different from TCP filter chain's.
 		currentListenerEntry.listener.FilterChains = mergeTCPFilterChains(mutable.Listener.FilterChains, listenerOpts, listenerMapKey, listenerMap)
+		currentListenerEntry.listener.ListenerFilters = mergeListenerFilters(currentListenerEntry.listener.ListenerFilters, mutable.Listener.ListenerFilters)
 	case TCPOverAuto:
 		// Merge two TCP filter chains. HTTP filter chain will not conflict with TCP filter chain because HTTP filter chain match for
 		// HTTP filter chain is different from TCP filter chain's.
 		currentListenerEntry.listener.FilterChains = mergeTCPFilterChains(mutable.Listener.FilterChains, listenerOpts, listenerMapKey, listenerMap)
+		currentListenerEntry.listener.ListenerFilters = mergeListenerFilters(currentListenerEntry.listener.ListenerFilters, mutable.Listener.ListenerFilters)
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/listener.go:1018]: Can you elaborate on why do it here? I think  for auto, we have already added inspector filter for it.

File: pilot/pkg/networking/core/v1alpha3/listener.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/listener.go:1545]: I mean you just return the combination of a and b  for `TLSInspector` and `HTTPInspector`.

So how about b contains another listener filter that a does not contain?

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 4]
```diff
@@ -984,6 +984,7 @@ func (lb *ListenerBuilder) buildSidecarOutboundListenerForPortOrUDS(listenerOpts
 		if currentListenerEntry != nil {
 			currentListenerEntry.listener.FilterChains = mergeTCPFilterChains(mutable.Listener.FilterChains,
 				listenerOpts, listenerMapKey, listenerMap)
+			currentListenerEntry.listener.ListenerFilters = mergeListenerFilters(currentListenerEntry.listener.ListenerFilters, mutable.Listener.ListenerFilters)
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/listener.go:987]: Can you elaborate here why need mergeListenerFilters for NoConflict

**Comment by howardjohn** [comment]
@hzxuzhonghu I modified the PR and test to be generic regardless of conflict type - we always take the union of the filters when we merge. The test also tests all possible protocol combinations. PTAL.

Note: the behavior is the same as original PR, just simpler implementation

**Review by hzxuzhonghu — APPROVED** [review]
This is better

**Pull request merged**